interfaces.go

```diff
@@ -1,46 +1,96 @@
+	// performing a topological sort.
+	ListNodesWithoutInboundConnections() map[string]Node[NodeType]
+	// Clone creates an independent copy of the current directed graph.
+	Clone() DirectedGraph[NodeType]
+	// HasCycles performs cycle detection and returns true if the DirectedGraph has cycles.
+	HasCycles() bool
+	// PopReadyNodes lists all nodes that have finalized their status, whether resolved
+	// or unresolvable, and clears the list.
+	PopReadyNodes() []*node[NodeType]
+	// PushStartingNodes pushes the initial nodes without dependencies for access with PopReadyNodes()
+	// O(N), so use only at the start.
+	PushStartingNodes() error
 
-    // Mermaid outputs the graph as a Mermaid string.
-    Mermaid() string
+	// Mermaid outputs the graph as a Mermaid string.
+	Mermaid() string
 }
 
 // Node is a single point in a DirectedGraph.
 type Node[NodeType any] interface {
-    // ID returns the unique identifier of the node in the DG.
-    ID() string
-    // Item returns the underlying item for the node.
-    Item() NodeType
-    // Connect creates a new connection from the current node to the specified node. If the specified node does not
-    // exist, ErrNodeNotFound is returned. If the connection had created a cycle, ErrConnectionWouldCreateACycle
-    // is returned.
-    Connect(toNodeID string) error
-    // DisconnectInbound removes an incoming connection from the specified node. If the connection does not exist, an
-    // ErrConnectionDoesNotExist is returned.
-    DisconnectInbound(fromNodeID string) error
-    // DisconnectOutbound removes an outgoing connection to the specified node. If the connection does not exist, an
-    // ErrConnectionDoesNotExist is returned.
-    DisconnectOutbound(toNodeID string) error
-    // Remove removes the current node and all connections from the DirectedGraph.
-    Remove() error
-    // ListInboundConnections lists all inbound connections to this node.
-    ListInboundConnections() (map[string]Node[NodeType], error)
-    // ListOutboundConnections lists all outbound connections from this node.
-    ListOutboundConnections() (map[string]Node[NodeType], error)
+	// ID returns the unique identifier of the node in the DG.
+	ID() string
+	// Item returns the underlying item for the node.
+	Item() NodeType
+	// Connect creates a new connection from the current node to the specified node. If the specified node does not
+	// exist, ErrNodeNotFound is returned. If the connection had created a cycle, ErrConnectionWouldCreateACycle
+	// is returned. Sets the dependency type to an AND dependency.
+	Connect(toNodeID string) error
+	// ConnectDependency creates a new connection from the specified node to the current node.
+	// The dependency type is set to determine when the node becomes finalized.
+	// If the specified node does not exist, ErrNodeNotFound is returned. If the connection had
+	// created a cycle, ErrConnectionWouldCreateACycle is returned.
```

Comment:
s/had/would/

Ditto line 67.

Reply:
~~Sure. Changed. The IDE didn't like the conditional clause, but in this case we're testing for the possibility. We're not detecting it after the fact.~~

I just noticed that it doesn't check for cycles in the current implementation. It's more efficient to only do that once all connections are made with the separate `HasCycles()` call. So I updated the comments to reflect the actual errors returned.